ppa/archive/management/commands/generate_textcorpus.py

```diff
@@ -2,199 +2,238 @@
-
-        # if pages, save json
-        if pages:
-            filename = os.path.join(self.path_texts, meta[self.WORK_ID_FIELD]+'.json')
-            os.makedirs(self.path_texts,exist_ok=True)
-            with open(filename,'w') as of:
-                json.dump(pages, of, indent=2)
-        
-        # otherwise, returned filename is blank to indicate no file saved
-        else:
-            filename=''
-        
-        return filename,meta
+        # add --doc-limit argument (determines how many results to retrieve from solr)
+        parser.add_argument(
+            "--doc-limit",
+            type=int,
+            default=-1,
+            help="Limit on the number of documents for corpus generation."
+            "The default of -1 considers ALL documents.",
+        )
 
+        # add --batch-size argument (for solr querying)
+        parser.add_argument(
+            "--batch-size",
+            type=int,
+            default=DEFAULT_BATCH_SIZE,
+            help="Number of docs to query from solr at one time",
+        )
 
-    def save(self):
-        """Save the generated corpus text and metadata to files on disk"""
+        # add --dry-run argument (don't save anything, just iterate)
+        parser.add_argument(
+            "--dry-run",
+            action="store_true",
+            help="Do not save anything, just iterate over solr",
+        )
 
-        # save docs and gather metadata
-        metadata=[]
-        pdesc='Saved text to'
-        pbar=tqdm(total=self.doc_count, desc=f'{pdesc}: ...')
-        for doc_id in self.doc_ids:
-            # get saved filename and found metadata for this document
-            fn,meta = self._save_doc(doc_id)
+    #### SOLR ####
 
-            # if we saved, update progress bar desc
-            if fn: pbar.set_description(f'{pdesc}: {fn}')
+    @cached_property
+    def query_set(self):
+        """
+        A cached representation of the query set;
+        it discourages opening multiple connections to database.
+        """
+        return SolrQuerySet()
 
-            # tick
-            pbar.update()
+    def iter_solr(self, item_type="page"):
+        """
+        Iterate over solr documents of a certain `item_type`.
+        """
 
-            # add this doc's meta to metadata
-            metadata.append(meta)
-        pbar.close()
+        # get params
+        batch_size, lim = self.batch_size, self.doclimit
 
-        # save metadata csv
-        with open(self.path_metadata,'w') as of:
-            json.dump(metadata, of, indent=2)
-        print(f'Saved metadata to: {self.path_metadata}')
+        # search for this item type, order by id
+        qset = self.query_set.search(item_type=item_type).order_by("id")
 
+        # if we're looking for pages, rename those fields
+        if item_type == "page":
+            qset = qset.only(*self.PAGE_FIELDLIST)
+        elif item_type == "work":
+            qset = qset.only(*self.WORK_FIELDLIST)
```

Comment:
yes! I've been meaning to implement this. what is the field to query on there? I didn't see it in the solr original fields:

```
 "_version_": 1757837811292045312,
   "author": "Sheridan, Thomas, 1719-1788",
   "author_exact": "Sheridan, Thomas, 1719-1788",
   "cluster_id_s": "sheridancompleteSV",
   "collections": [
     "Dictionaries",
     "Linguistic"
   ],
   "collections_exact": [
     "Dictionaries",
     "Linguistic"
   ],
   "group_id_s": "CB0126082904",
   "id": "CB0126082904",
   "item_type": "work",
   "last_modified": "2023-02-14T19:59:33.011Z",
   "order": 0,
   "pub_date": 1796,
   "pub_place": "Philadelphia ",
   "publisher": "W. Young, Mills & Son, booksellers and stationers, no. 52, Second-Street, corner of Chesnut-Street",
   "sort_title": "complete dictionary of the English language, both with regard to sound and meaning: one main object of which is, to establish a plain and permanent standard of pronunciation. To which is prefixed, a prosodial grammar. By Thomas Sheridan, A.M.",
   "source_id": "CB0126082904",
   "source_t": [
     "Gale"
   ],
   "source_url": "https://link.gale.com/apps/doc/CB0126082904/ECCO?u=prin77918&sid=gale_api&xid=a5b78388",
   "subtitle": "both with regard to sound and meaning: one main object of which is, to establish a plain and permanent standard of pronunciation. To which is prefixed, a prosodial grammar. By Thomas Sheridan, A.M.",
   "title": "A complete dictionary of the English language,",
   "work_type_s": "full-work"
```

Reply:
wow, the danger of trusting my memory. I just checked, you don't need to filter in solr because it isn't there! when a work is suppressed we index a stub record (id only) to make sure we replace any existing data. Sorry for misleading!

I think this does make the check on totals more important, though (# unsuppressed works, # pages expected in the db, actuals in the Solr index). I think I've done something like that elsewhere and can dig it up so we can adapt.

Reply:
Just confirming: neither suppressed works nor the pages of suppressed works are in solr?

Reply:
Correct, that is what we expect (as long as indexing is in sync with the db). Want to make a note in the comments in the code as a reminder?